### PR TITLE
Experimental: Partitioned Permutations

### PR DESCRIPTION
**Pull request by BjSchaefer**
This is a new module for working with partitioned permutations which are used in Free Probablility Theory. For more information see the Readme file.

The module uses the module SetPartitions which is currently in the folder experimentals. To compile properly, we needed to rename the project folder to ZPartitionedPermutations so that this folder is below the folder SetPartitions within the folder experimentals.

### CODE REVIEW

**Comment by benlorenz** [comment]
> The module uses the module SetPartitions which is currently in the folder experimentals. To compile properly, we needed to rename the project folder to ZPartitionedPermutations so that this folder is below the folder SetPartitions within the folder experimentals.

Please add both folders to `orderedpkgs` instead:
https://github.com/oscar-system/Oscar.jl/blob/222c2000987befa84b9376a235ae54fd1e1d8f44/experimental/Experimental.jl#L15-L22

**Comment by lgoettgens** [comment]
Thanks for your contribution! I have not much background in the mathematics involved, but will point out some general coding things. 

> The module uses the module SetPartitions which is currently in the folder experimentals. To compile properly, we needed to rename the project folder to ZPartitionedPermutations so that this folder is below the folder SetPartitions within the folder experimental.

Please add both folders in 
https://github.com/oscar-system/Oscar.jl/blob/222c2000987befa84b9376a235ae54fd1e1d8f44/experimental/Experimental.jl#L19
together with a comment like the one already existent to specify the order of inclusion.

**Review by lkastner — DISMISSED** [review]
I'll block merging until the comment of @benlorenz has been addressed. There is a dedicated mechanism for loading certain packages in an ordered fashion. The side effect is that if a dependency that is in experimental gets discarded, then so will any package depending thereupon.

File: experimental/SymmetricIntersections/docs/doc.main
Comment (lgoettgens) [inline experimental/SymmetricIntersections/docs/doc.main:1]: why this change?

File: experimental/PartitionedPermutations/README.md
Comment (lgoettgens) [inline experimental/PartitionedPermutations/README.md]: Please add a website or something similar with contact details.

File: experimental/PartitionedPermutations/README.md
Comment (lgoettgens) [inline experimental/PartitionedPermutations/README.md]: please use citations using the bibliography. see [docs.oscar-system.org/dev/DeveloperDocumentation/documentation/#Updating-the-bibliography](https://docs.oscar-system.org/dev/DeveloperDocumentation/documentation/#Updating-the-bibliography)

File: experimental/PartitionedPermutations/src/EnumeratePartitionedPermutations.jl
Comment (lgoettgens) [inline experimental/PartitionedPermutations/src/EnumeratePartitionedPermutations.jl]: The `SymmetricGroup` from AbstractAlgebra should not be used in real usage code. Please use `symmetric_group` (backed by GAP) instead.
(@fingolfin please correct me if I am wrong)

File: experimental/PartitionedPermutations/src/PartitionedPermutation.jl
Comment (lgoettgens) [inline experimental/PartitionedPermutations/src/PartitionedPermutation.jl]:
```suggestion
```

never set

File: experimental/PartitionedPermutations/src/PartitionedPermutation.jl
Comment (lgoettgens) [inline experimental/PartitionedPermutations/src/PartitionedPermutation.jl]:
```suggestion
    function PartitionedPermutation(_p::Perm{Int}, _V::Vector{Int}; check::Bool=true)
```
`check` should be a kwarg.

Comment (lgoettgens) [inline experimental/PartitionedPermutations/src/PartitionedPermutation.jl]: In julia you don't need to distinguish parameter variables from other variables with a prefix `_`, I would rather call it bad style here

File: experimental/PartitionedPermutations/src/PartitionedPermutation.jl
Comment (lgoettgens) [inline experimental/PartitionedPermutations/src/PartitionedPermutation.jl]:
```suggestion
function partitioned_permutation(p::Perm{Int}, V::Vector{Int}; check::Bool=true)
    return PartitionedPermutation(p, V; check)
end
```
kwargs

File: experimental/PartitionedPermutations/src/PartitionedPermutation.jl
Comment (lgoettgens) [inline experimental/PartitionedPermutations/src/PartitionedPermutation.jl]: access to `n` is unsafe. use an accessor function instead.

File: experimental/PartitionedPermutations/src/PartitionedPermutation.jl
Comment (lgoettgens) [inline experimental/PartitionedPermutations/src/PartitionedPermutation.jl]: CITE missing

**Comment by thofma** [comment]
Since this adds functions for `SetPartition` from experimental/SetPartition, is there a particular reason to not combine those too? Maybe with some generic "FreeProbabilityTheory" name? Or alternatively, would it be an option to move those methods to the SetPartition module?

File: experimental/PartitionedPermutations/src/PartitionedPermutation.jl
Comment (lgoettgens) [inline experimental/PartitionedPermutations/src/PartitionedPermutation.jl]:
```suggestion
julia> length(partitioned_permutation(Perm([2, 1]), [1, 1]))
```
use the lowercase constructor in examples (also at other places)

File: experimental/PartitionedPermutations/src/SetPartitionFunctions.jl
Comment (lgoettgens) [inline experimental/PartitionedPermutations/src/SetPartitionFunctions.jl]: This function seems to belong to the `SetPartition` module instead.

File: experimental/PartitionedPermutations/src/SetPartitionFunctions.jl
Comment (lgoettgens) [inline experimental/PartitionedPermutations/src/SetPartitionFunctions.jl]: Overloading `<=` seems dangerous and unexpected to me. Something like `is_dominated_by` seems more suitable to me

This function seems to belong to the `SetPartition` module instead.

File: experimental/PartitionedPermutations/src/SetPartitionFunctions.jl
Comment (lgoettgens) [inline experimental/PartitionedPermutations/src/SetPartitionFunctions.jl]: unsafe access to internal attributes

File: experimental/PartitionedPermutations/src/SetPartitionFunctions.jl
Comment (lgoettgens) [inline experimental/PartitionedPermutations/src/SetPartitionFunctions.jl]:
```suggestion
        if haskey(block_map, block) && W_vec[index] != block_map[block]
```
`if`s need no parentheses

File: experimental/PartitionedPermutations/src/SetPartitionFunctions.jl
Comment (lgoettgens) [inline experimental/PartitionedPermutations/src/SetPartitionFunctions.jl]: This function seems to belong to the `SetPartition` module instead.

File: experimental/PartitionedPermutations/src/SetPartitionFunctions.jl
Comment (lgoettgens) [inline experimental/PartitionedPermutations/src/SetPartitionFunctions.jl]: please add a description/definition to the docstring. This contains zero additional information compared to the signature alone

**Comment by BjSchaefer** [comment]
The SetPartitions do not exactly contribute to Free Probability Theory but rather to certain compact quantum groups. However, it is very much an option to move the functions for SetPartitions to the corresponding module.

**Comment by thofma** [comment]
> However, it is very much an option to move the functions for SetPartitions to the corresponding module.

I think that would be great.

> The SetPartitions do not exactly contribute to Free Probability Theory but rather to certain compact quantum groups.

OK, that makes sense. I think what we want to avoid is having too many separated (but interdependent) modules in experimental. So if there are more things incoming in the same "broad" area (it could be "Free probability theory and quantum groups"), at some point we can and probably should think about consolidating things. Makes it also easier to move things from experimental to src (which should be the endgame).

Edit: These are just some things to keep in mind, nothing that should hold up this PR here.

File: experimental/PartitionedPermutations/README.md
Comment (BjSchaefer) [reply experimental/PartitionedPermutations/README.md]: has been addressed

File: experimental/PartitionedPermutations/src/PartitionedPermutation.jl
Comment (BjSchaefer) [reply experimental/PartitionedPermutations/src/PartitionedPermutation.jl]: has been addressed

File: experimental/SymmetricIntersections/docs/doc.main
Comment (BjSchaefer) [reply experimental/SymmetricIntersections/docs/doc.main:1]: It seems that there is no change for the file file is empty in both branches "master" and "bs/partitioned_permutation". It is not clear why Git says that there has been a change.

Comment (thofma) [reply experimental/SymmetricIntersections/docs/doc.main:1]: You changed the permissions of the file. I would suggest to undo this.

File: experimental/PartitionedPermutations/src/PartitionedPermutation.jl
Comment (thofma) [inline experimental/PartitionedPermutations/src/PartitionedPermutation.jl]: I would suggest to come up with a more descriptive name than `length2`. Maybe `length_adjusted` or `adjusted_length`? Alternatively keep it internal and don't export it.

File: experimental/PartitionedPermutations/src/EnumeratePartitionedPermutations.jl
Comment (thofma) [inline experimental/PartitionedPermutations/src/EnumeratePartitionedPermutations.jl]:
```suggestion
function enumerate_partitioned_permutations(n::Int)
```
For consistency with the other exported `*permutation` function.

File: experimental/PartitionedPermutations/src/PartitionedPermutation.jl
Comment (lgoettgens) [reply experimental/PartitionedPermutations/src/PartitionedPermutation.jl]: Can you click on "resolve conversation" for everything that you adressed? That makes it easier to keep an overview what's left to do

File: docs/oscar_references.bib
Comment (lgoettgens) [inline docs/oscar_references.bib]:
```suggestion
	title = {Second order freeness and fluctuations of random matrices. III: Higher order freeness and free cumulants},
```
The bibtex Parser does not like braces, but instead keeps capitalization by itself.

Comment (lgoettgens) [reply docs/oscar_references.bib]: And there are more bibtex issues (see the bibtool test ci job for details)

Comment (fingolfin) [reply docs/oscar_references.bib]: see also https://docs.oscar-system.org/dev/DeveloperDocumentation/documentation/#Updating-the-bibliography

File: experimental/PartitionedPermutations/src/PartitionedPermutations.jl
Comment (lgoettgens) [inline experimental/PartitionedPermutations/src/PartitionedPermutations.jl]: A _ prefix denotes an internal function. Thus, this shouldn't get exported.

File: experimental/PartitionedPermutations/src/EnumeratePartitionedPermutations.jl
Comment (BjSchaefer) [reply experimental/PartitionedPermutations/src/EnumeratePartitionedPermutations.jl]: It seems that we cannot use symmetric_group instead of Generic.SymmetricGroup. Indeed, the former yields an object of type `PermGroup` which is not accepted by the function `Generic.elements!`.

Comment (joschmitt) [reply experimental/PartitionedPermutations/src/EnumeratePartitionedPermutations.jl]: You can just do `for p in symmetric_group(n)`.

File: experimental/SetPartitions/src/SetPartition.jl
Comment (lgoettgens) [inline experimental/SetPartitions/src/SetPartition.jl]:
```suggestion
     maximum(vec; init=0)
```

File: experimental/SetPartitions/src/SetPartition.jl
Comment (lgoettgens) [inline experimental/SetPartitions/src/SetPartition.jl]:
```suggestion
julia> number_of_blocks(set_partition([1, 2, 3], [2, 1, 3, 3]))
```
use the lowercase constructor in doctests

File: experimental/SetPartitions/src/SetPartition.jl
Comment (lgoettgens) [inline experimental/SetPartitions/src/SetPartition.jl]:
```suggestion
    number_of_blocks(p::SetPartition)

Return the number of blocks of `p`.
```
consistency with other docstrings in this file

File: experimental/SetPartitions/src/SetPartition.jl
Comment (lgoettgens) [inline experimental/SetPartitions/src/SetPartition.jl]:
```suggestion
julia> is_dominated_by(set_partition([1, 1, 2], [1, 2, 3]), set_partition([1, 1, 2], [1, 2, 1]))
```
use the lowercase constructor in doctests

File: experimental/PartitionedPermutations/src/EnumeratePartitionedPermutations.jl
Comment (lgoettgens) [inline experimental/PartitionedPermutations/src/EnumeratePartitionedPermutations.jl]: `_enumerate_partitioned_perm` does not exist anymore

File: experimental/PartitionedPermutations/src/PartitionedPermutation.jl
Comment (lgoettgens) [inline experimental/PartitionedPermutations/src/PartitionedPermutation.jl]: This name is rather unfortunate given that `set_partition` is not the corresponding setter but does something completely different.
Furthermore, in the usual julia style, one would name this function `partition(::PartitionedPermutation)`

File: experimental/PartitionedPermutations/src/PartitionedPermutationProducts.jl
Comment (lgoettgens) [inline experimental/PartitionedPermutations/src/PartitionedPermutationProducts.jl]: described where?

File: experimental/PartitionedPermutations/src/PartitionedPermutations.jl
Comment (lgoettgens) [inline experimental/PartitionedPermutations/src/PartitionedPermutations.jl]:
```suggestion
```
The need to import `Generic` is basically always a sign that you use something that is not intended to be used directly. (like `Perm` in this case)

File: experimental/PartitionedPermutations/src/EnumeratePartitionedPermutations.jl
Comment (BjSchaefer) [reply experimental/PartitionedPermutations/src/EnumeratePartitionedPermutations.jl]: Well, then I get a problem later because elements of `symmetric_group(n)` are of type `PermGroupElem` and not of type `Perm`. Is it generally recommended to use the types `PermGroup` and `PermGroupElem` instead of `Generic.SymmetricGroup` and `Perm`, respectively?

Comment (lgoettgens) [reply experimental/PartitionedPermutations/src/EnumeratePartitionedPermutations.jl]: Yes, the `AbstractAlgebra.Generic.SymmetricGroup` is only a toy-like implementation that is not integrated into other parts of OSCAR. Thus, they should not be used outside of AbstractAlgebra.jl

Comment (joschmitt) [reply experimental/PartitionedPermutations/src/EnumeratePartitionedPermutations.jl]: If functionality for `PermGroup` or `PermGroupElem` should be missing, let us know and somebody can add it.

Comment (BjSchaefer) [reply experimental/PartitionedPermutations/src/EnumeratePartitionedPermutations.jl]: Here is a functionality that seems to be missing: We need a function like `cycles(g::Perm)` that returns a list of the cycles of the permutation `g`. Apparently, one can only obtain the cycle structure of a `PermGroupElem` with `cycle_structure(g::PermGroupElem)` but not the cycles explicitely.

Comment (joschmitt) [reply experimental/PartitionedPermutations/src/EnumeratePartitionedPermutations.jl]: @fingolfin @ThomasBreuer could we add `cycles` for `PermGroupElem`?
~Easiest would be to copy the code from `cycles` and call it on `Vector(g)` I guess.~

Comment (joschmitt) [reply experimental/PartitionedPermutations/src/EnumeratePartitionedPermutations.jl]: `cycles` is now available for `PermGroupElem` (if you merge `master` into this branch). Sorry for the delay.

Comment (BjSchaefer) [reply experimental/PartitionedPermutations/src/EnumeratePartitionedPermutations.jl]: Thank you very much! We have now replaced every mention of Perm with PermGrouElem.

**Comment by joschmitt** [comment]
@BjSchaefer Please mark the pull request as "ready for review" when you are done. I assume this is the case, but I did not follow all of the discussion, so I am not sure.

**Review by joschmitt — APPROVED** [review]
Thank you!
I don't know about the mathematics, but for `experimental` this certainly looks good enough.
Whoever has write access to master: this can be merged from my point of view.

File: experimental/Experimental.jl
Comment (lgoettgens) [inline experimental/Experimental.jl]:
```suggestion
  "BasisLieHighestWeight",   # needs code from LieAlgebras
```

File: experimental/PartitionedPermutations/src/PartitionedPermutations.jl
Comment (lgoettgens) [inline experimental/PartitionedPermutations/src/PartitionedPermutations.jl]:
```suggestion
```
these are already contained in `src/exports.jl` and need not be duplicated

File: experimental/PartitionedPermutations/src/PartitionedPermutationProducts.jl
Comment (lgoettgens) [inline experimental/PartitionedPermutations/src/PartitionedPermutationProducts.jl]: Maybe rename to `factorization` or `factorize` in a future PR. (no change needed for this one, just some thought)

**Review by lgoettgens — APPROVED** [review]
Two small nitpicks, otherwise no objections from my side. But as Johannes, I am not familiar with the mathematics involved.

File: experimental/PartitionedPermutations/src/PartitionedPermutationProducts.jl
Comment (BjSchaefer) [reply experimental/PartitionedPermutations/src/PartitionedPermutationProducts.jl]: This seems to me a very good suggestion, so I will change the name right now

Comment (lgoettgens) [reply experimental/PartitionedPermutations/src/PartitionedPermutationProducts.jl]: Sorry for not thinking this through before posting. `factorization`(s) seems to already exist in the Oscar universe while `factorize` does not. So `factorization` should be the preferred name I think.

Comment (BjSchaefer) [reply experimental/PartitionedPermutations/src/PartitionedPermutationProducts.jl]: Done

**Comment by fieker** [comment]
On Tue, Apr 09, 2024 at 05:33:21AM -0700, Lars Göttgens wrote:
> @lgoettgens commented on this pull request.
> 
> 
> 
> > +            upper_points(cycle_partition(perm(symmetric_group(length(pp_1)), 1:length(pp_1)))))
> +    end
> +end
> +
> +"""
> +    factorization_partitioned_permutation(pp::PartitionedPermutation)
> +
> +Return the factorization of `pp` in form of a set of 2-tuples.
> +
> +# Examples
> +```jldoctest
> +julia> length(factorization_partitioned_permutation(partitioned_permutation(perm(symmetric_group(3), [2, 1, 3]), [1, 1, 2])))
> +2
> +```
> +"""
> +function factorization_partitioned_permutation(pp::PartitionedPermutation)
> 
> Sorry for not thinking this through before posting. `factorization`(s) seems to already exist in the Oscar universe while `factorize` does not. So `factorization` should be the preferred name I think.

factor is the usual Oscar name to obtain a factorisation.
factorisations is for the special case of algebraic integers where all
factorisations into irreducibles can be computed, but a single
factorisation should "factor"

I'd actually prefer:

function factor(pp::PartitionedPermutation)

as the signature - unless there is a reason to repeat the paritioned
permutation bit in the name and the signature

**Comment by lgoettgens** [comment]
> On Tue, Apr 09, 2024 at 05:33:21AM -0700, Lars Göttgens wrote: @lgoettgens commented on this pull request. > + upper_points(cycle_partition(perm(symmetric_group(length(pp_1)), 1:length(pp_1))))) + end +end + +""" + factorization_partitioned_permutation(pp::PartitionedPermutation) + +Return the factorization of `pp` in form of a set of 2-tuples. + +# Examples +```jldoctest +julia> length(factorization_partitioned_permutation(partitioned_permutation(perm(symmetric_group(3), [2, 1, 3]), [1, 1, 2]))) +2 +``` +""" +function factorization_partitioned_permutation(pp::PartitionedPermutation) Sorry for not thinking this through before posting. `factorization`(s) seems to already exist in the Oscar universe while `factorize` does not. So `factorization` should be the preferred name I think.
> factor is the usual Oscar name to obtain a factorisation. factorisations is for the special case of algebraic integers where all factorisations into irreducibles can be computed, but a single factorisation should "factor" I'd actually prefer: function factor(pp::PartitionedPermutation) as the signature - unless there is a reason to repeat the paritioned permutation bit in the name and the signature
> […](#)
> -- Reply to this email directly or view it on GitHub: [#3439 (comment)](https://github.com/oscar-system/Oscar.jl/pull/3439#discussion_r1557556876) You are receiving this because you are subscribed to this thread. Message ID: ***@***.***>

Thanks claus! I didn't remember this one.

File: experimental/PartitionedPermutations/src/PartitionedPermutationProducts.jl
Comment (BjSchaefer) [reply experimental/PartitionedPermutations/src/PartitionedPermutationProducts.jl]: Okay, as 'factor' is now the favourite I will change the name again later.

Comment (lgoettgens) [reply experimental/PartitionedPermutations/src/PartitionedPermutationProducts.jl]: Thanks, and sorry again for the triple work here.

**Pull request merged**
